---
title: "Four key homologs of Gat201"
author: "Edward Wallace"
date: "07/12/2021"
output: 
  html_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
library(readr)
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)
library(cowplot)
library(Biostrings)
library(msa)
library(here)
```

# Summary 

This is an introductory homology analysis of _Cryptococcus neoformans_ Gat201, along with 3 other landmark homologs.

We aim to plot the domain structure, and to show a multiple sequence alignment of the shared GATA zinc-finger domain.

# Load data on interpro domains and sequences

```{r load_data}
read_interpro_domains <- function(accession,
                                  prefix = here::here("data","Interpro_domains_"), 
                                  suffix = ".tsv") {
  paste0(prefix, accession, suffix) %>%
    readr::read_tsv(col_names = c("domain_accession",	"domain_name", "domain_database"	, "domain_type", "integrated_into", "integrated_signatures", "go_terms", "protein_accession", "protein_length", "domain_matches"),
                    skip = 1) 
}
# read_interpro_domains("J9VWG0") %>% View()

protein_info <- 
  here::here("data","four_homolog_info.txt") %>%
  readr::read_tsv() %>%
  mutate(Species = factor(Species, levels=Species))

domain_info_all <- protein_info %>%
  dplyr::group_by(Uniprot) %>%
  dplyr::do(read_interpro_domains(.$Uniprot)) 

domain_info_nhrgata <- domain_info_all %>%
  dplyr::filter(domain_accession == "IPR013088") %>%
  dplyr::select(Uniprot, domain_accession, domain_name, domain_matches) %>%
  tidyr::separate(domain_matches, 
                  into = c("domain_start","domain_end"),
                  convert = TRUE) %>%
  dplyr::full_join(protein_info)
```

```{r print_domain_info_nhrgata}
domain_info_nhrgata
```


# Domain plot

```{r domain_pos_nhrgata_longdf, eval=FALSE, include = FALSE}
# trial code to move domain positions around so it's "easy" to plot
# ugly and for now abandoned
domain_pos_nhrgata_longdf <-
  domain_info_nhrgata %>%
  dplyr::select(Uniprot, Species, Length, domain_start, domain_end) %>%
  dplyr::mutate(Start = 1, pos_align = domain_start) %>%
  tidyr::pivot_longer(-c(Uniprot,Species,pos_align), values_to = "pos")
```


```{r domain_plot, fig.height = 2.5, fig.width = 5}

xvalue_to_expand_for_text <- with(domain_info_nhrgata,
                                max(Length - domain_start) + 20)

ggplot(data = domain_info_nhrgata) +
  theme_nothing() +
  theme(axis.text.y = element_text()) +
  geom_segment(aes(x = 1 - domain_start, xend = Length - domain_start,
                   y = Species, yend = Species), 
               size = 8, colour = "grey50") +
  geom_segment(aes(x = 1, xend = domain_end - domain_start,
                   y = Species, yend = Species), 
               size = 8, colour = "red2") +
  geom_text(aes(x = 1 - domain_start, y = Species, label = 1), 
            hjust = 1.1) +
  geom_text(aes(x = Length - domain_start, y = Species, label = Length), 
            hjust = -0.1)  +
  geom_text(aes(x = 1, y = Species, label = domain_start), 
            vjust = -1.3, colour = "darkred") +
  geom_text(aes(x = domain_end - domain_start, y = Species, label = domain_end), 
            vjust = -1.3, colour = "darkred") +
  expand_limits(x = xvalue_to_expand_for_text)
```

## Multiple sequence alignment of GATA domains

```{r load_sequences}
protein_seqs <- here::here("data", "four_homolog_sequences_uniprot.fasta") %>% 
  Biostrings::readAAStringSet()

names(protein_seqs) <-
  names(protein_seqs) %>%
  stringr::str_sub(1,6)

# protein_seqdf <-
#   tibble(Uniprot = names(protein_seqs) %>%
#            stringr::str_sub(1,6),
#            Sequence = as.character(protein_seqs)
#            )

# nhrgata_ranges <- IRanges(start = domain_info_nhrgata$domain_start,
#                           end = domain_info_nhrgata$domain_start)
# 
# nhrgata_seqs <- Biostrings::extractAt(protein_seqs,nhrgata_ranges)
```

```{r extract_nhrgata_seqs}
# domain_info_nhrgata_wseq <- 
#   domain_info_nhrgata %>%
#   # group_by(Uniprot) %>%
#   left_join(protein_seqdf) %>%
#   mutate(domain_seq = stringr::str_sub(Sequence, domain_start, domain_end)) %>%
#   select(-Sequence)
  
domain_seqs_nhrgata <- 
  AAStringSet( x = protein_seqs[domain_info_nhrgata$Uniprot],
               start = domain_info_nhrgata$domain_start,
               # end = domain_info_nhrgata$domain_end,
               width = 50)

domain_msa <- msa(domain_seqs_nhrgata,method = "Muscle")

domain_msa %>%
  as("AAStringSet") %>%
  writeXStringSet(file = here::here("results", "nhrgatadomain_muscle.fasta"))
```

```{r ggmsa_nhrgata,fig.height = 2, fig.width = 6}
domain_msa %>%  as("AAStringSet") %>% ggmsa::ggmsa()
```

## TO DO NEXT

- get sequence names and ordering consistent across figures
- make a composite figure
- decide what to say/do about disordered regions
